### ISSUE TITLE
Add support for multiple active subscriptions per user

### ISSUE THREAD

**Issue by mindctrl** [restrictcontentpro/restrict-content-pro#774]
We've been asked a lot about supporting multiple active subscriptions per user. We need to add support for it. There's a lot to consider for this, and we should make a master list here to refer to when the time comes.

[x] How it works with upgrades/downgrades
[x] Helper functions currently assume one sub. We'll need to introduce new helpers, and possibly update existing ones.
[x] Member status and how it's affected by the fact that one sub could be expired and one might not be.

More items surely to follow. Feel free to edit this comment to add to the list so we have one list.

**Comment by pippinsplugins** [comment]
We'll need to determine the data storage for sub IDs, expiration dates, statuses, etc. The current system with usermeta won't work well.

This would be a lot easier if coupled with a custom table.

We will also need to update the payments database so that each payment can be related to a specific subscription.

**Comment by greguly** [comment]
I am doing a hack for this, as a customer will be selling 'movie rentals' and my solution would be hooking into payments database.

Please let me know if you would like help.

**Comment by nosegraze** [comment]
Ideas for columns in a subscriptions table from our meeting:

- Subscription ID
- Subscription level ID
- Member ID
- Initial amount
- Recurring amount
- Discount code applied (maybe)
- Trialling until (or something else to indicate if a trial was used)
- Start date
- Expiration date
- Status
- Payment profile ID

Feel free to add any I may have missed.

**Comment by pippinsplugins** [comment]
Added `gateway` to the list.

**Comment by pippinsplugins** [comment]
New request https://secure.helpscout.net/conversation/401738526/63523/?folderId=841776

**Comment by pippinsplugins** [comment]
New request just came through live chat for this.

**Comment by brandomeniconi** [comment]
We also need this, we are forced to move to another plugin if this doesn't land soon :(.

**Comment by pippinsplugins** [comment]
@brandomeniconi this (and a number of changes that are required for this to be completed) is in progress. We hope to have updates on it soon but there's not an exact ETA.

**Comment by mindctrl** [comment]
Here's the proposed schema for subscriptions. Some of these column types and sizes are the same as some from payments, but it's a good time to evaluate whether we should change any.

`object_id` is indexed. We'll be querying by this to determine a member's access to content.

`object_type` is not indexed at the moment, but in the future we'll probably want it to be so we can support subscriptions for things other than RCP levels - such as posts, pages, etc. We could just do it now.

`member_id` is indexed, as we'll be using it to fetch member subscriptions all the time. Maybe this should be `user_id` for consistency?

`status` is indexed the same as the payments table.


Field | Type | Null | Key | Default | Extra
-- | -- | -- | -- | -- | --
id | bigint(20) unsigned | NO | PRI | NULL | auto_increment
object_id | bigint(20) unsigned | NO | MUL | NULL |
object_type | varchar(20) | NO |   | 'subscription' | 
member_id | mediumint(9) | NO | MUL | NULL |
initial_amount | mediumint(9) | NO | | NULL |
recurring_amount | mediumint(9) | NO | | NULL |
start_date | datetime | NO | | CURRENT_TIMESTAMP |
trial_end_date | datetime | YES | | NULL |
cancellation_date | datetime | YES | | NULL |
expiration_date | datetime | YES | | NULL |
status | varchar(12) | NO | MUL | active |
payment_profile_id | tinytext | YES | | NULL |
gateway | tinytext | NO | | NULL |


Thoughts?

**Comment by pippinsplugins** [comment]
Is `start_date` the creation date or the date upon which a subscriber is granted access? While we do not support future-dated subscriptions, I can see that being a possibility in the future. If it's meant to be the creation date, I think we should change that to `date_created`.

I feel `member_id` should be changed to `user_id`.

**Comment by nosegraze** [comment]
This is looking good! What about a column for `times_billed` ? This could come in handy for https://github.com/restrictcontentpro/restrict-content-pro/issues/766 where we may want to cancel the subscription after a certain number of times billed.

**Comment by mindctrl** [comment]
@pippinsplugins @nosegraze agreed on those.

Updated schema below. I changed:
-  `start_date` to `created_date` (keeping date at the end of the word for consistency with other date field names).
- `id` from bigint to int
- `member_id` to `user_id` and from `mediumint` to `int unsigned`
- added `times_billed` with default value of `1`

Field | Type | Null | Key | Default | Extra
-- | -- | -- | -- | -- | --
id | int unsigned | NO | PRI | NULL | auto_increment
object_id | bigint unsigned | NO | MUL | NULL |
object_type | varchar(20) | NO |   | 'subscription' | 
user_id | int unsigned | NO | MUL | NULL |
initial_amount | mediumint | NO | | NULL |
recurring_amount | mediumint unsigned | NO | | NULL |
created_date | datetime | NO | | CURRENT_TIMESTAMP |
trial_end_date | datetime | YES | | NULL |
cancellation_date | datetime | YES | | NULL |
expiration_date | datetime | YES | | NULL |
times_billed | smallint | NO | NO | 1 | |
status | varchar(12) | NO | MUL | active |
payment_profile_id | tinytext | YES | | NULL |
gateway | tinytext | NO | | NULL |

**Comment by pippinsplugins** [comment]
For `times_billed` I'm going to suggest that the initial signup, whether it is free, paid, or using a free trial be counted as `1`.

**Comment by pippinsplugins** [comment]
`initial_amount` and `recurring_amount` cannot be `mediumint` unless we're only permitting whole dollar amounts ;)

**Comment by tnorthcutt** [comment]
Wouldn’t you want to store cents, instead of dollars? That’s what eg Stripe does.

**Comment by pippinsplugins** [comment]
Storing in cents still needs support for decimal places if we're to support currencies like BTC that use 8 decimal places.

**Comment by mindctrl** [comment]
I'd **love** to store currency as an integer in cents, and use a multiplier on the way out to format it in the selected currency, though I didn't mean to do `int` in this case. Moving from floats to ints is a bigger discussion I'd love to have in the future!

On that note, should we have a `currency` column?

**Comment by pippinsplugins** [comment]
Yes, I think we should include a currency column, even though we'll likely not use it ourselves, enabling multi-currency support through extensions is a good idea.

**Comment by mindctrl** [comment]
Updated with `currency`.

Field | Type | Null | Key | Default | Extra
-- | -- | -- | -- | -- | --
id | int unsigned | NO | PRI | NULL | auto_increment
object_id | bigint unsigned | NO | MUL | NULL |
object_type | varchar(20) | NO |   | 'subscription' | 
user_id | int unsigned | NO | MUL | NULL |
currency | varchar(20) | NO | | USD |
initial_amount | mediumint | NO | | NULL |
recurring_amount | mediumint unsigned | NO | | NULL |
created_date | datetime | NO | | CURRENT_TIMESTAMP |
trial_end_date | datetime | YES | | NULL |
cancellation_date | datetime | YES | | NULL |
expiration_date | datetime | YES | | NULL |
times_billed unsigned | smallint | NO | NO | 1 | |
status | varchar(12) | NO | MUL | active |
payment_profile_id | tinytext | YES | | NULL |
gateway | tinytext | NO | | NULL |

**Comment by nosegraze** [comment]
One thing I'd really like to do during the migration process is make sure all the expiration dates are in the correct MySQL date format. An issue with CSV User Import https://github.com/restrictcontentpro/rcp-csv-user-import/issues/21 made it so imported users had a saved expiration date in whatever format it was in the CSV file. So it could actually be in the database as `17 Jan 2018`, for example. This doesn't cause any problems with restrictions, but it does mean the cron job for sending reminder notices doesn't work on those users since they don't get included in the date query for getting relevant members.

**Comment by mindctrl** [comment]
I created a separate issue for the custom table here: https://github.com/restrictcontentpro/restrict-content-pro/issues/1711

**Comment by tnorthcutt** [comment]
Added to original post:
> Reporting: if we add per-subscription level signup reports, we'll need to revisit when adding multiple sub support (https://github.com/restrictcontentpro/restrict-content-pro/issues/1645)

**Comment by lisag** [comment]
New comment on a blog post about this feature request: 

"I have been using RCP for a few year. Excellent product and the developer is a genius but the product has severe limitations. For example a member cannot be assigned two different subscriptions. This forces the developer to work only with “progressive” plans but not with distinct plans. I think RCP value would be greatly unchained with the feature of multiple subscriptions for members. Thank you."
